feat: deploy contract account

enitrat opened the pull request:
<!--- Please provide a general summary of your changes in the title above -->

## Pull Request type

<!-- Please try to limit your pull request to one type; submit multiple
pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying,
or link to a relevant issue. -->

Resolves: #422

## What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

- Moved `ContractAccountTrait` to be with `EOATrait` inside `evm` crate
- Refactored `ContractAccountTrait/EOATrait::at` to return `Option<Self>` depending on whether there is / the account at address `evm_address` is of the corresponding type
- `ContractAccountTrait::deploy` as the implementation logic
- `deploy_contract_account` internal KakarotCore method

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this does introduce a breaking change, please describe the impact and
migration path for existing applications below. -->

Eikix commented on crates/contracts/src/kakarot_core/kakarot.cairo:
```diff
@@ -78,6 +78,12 @@ mod KakarotCore {
+    #[derive(Drop, starknet::Event)]
+    struct ContractAccountDeployed {
+        #[key]
+        evm_address: EthAddress,
```
just thinking out loud, should we include the deployer (for ease of indexing)

Eikix commented on crates/evm/src/model/contract_account.cairo:
```diff
@@ -43,6 +82,20 @@ impl ContractAccountImpl of ContractAccountTrait {
 
+    /// Sets the nonce of a contract account.
+    /// The new nonce is written in Kakarot Core's contract storage.
+    /// The storage address used is h(sn_keccak("contract_account_nonce"), evm_address), where `h` is the poseidon hash function.
```
we should probably have a markdown document that records / centralizes all the storage addresses we decide to use

Eikix commented on crates/evm/src/model/contract_account.cairo:
```diff
@@ -24,12 +27,48 @@ struct ContractAccount {
+    /// # Errors
+    /// Returns an `EVMError` if the nonce of the contract account for
+    /// this EVM address was not `0`.
+    fn deploy(evm_address: EthAddress) -> Result<ContractAccount, EVMError> {
```
I think we need to add a todo for:
- setting code
- setting balance (in case of a payable constructor)

Eikix replied:
Nvm it's down in kakarot_core above,
Then maybe this function should be call initialize?
or "touch"?

Eikix replied:
okay

Eikix commented on crates/contracts/src/kakarot_core/kakarot.cairo:
```diff
@@ -295,5 +301,38 @@ mod KakarotCore {
+                gas_limit: 0,
+            );
+            //TODO gas params
+            if execution_result.status == Status::Reverted {
```
In this case, should the ContractAccount be "undeployed"? i.e. the nonce be set back to 0?
What does the spec say

enitrat replied:
If the context reverts then pending storage updates will not be committed so it should be a problem

Eikix replied:
but the nonce will be written to 1 above

enitrat replied:
Yes because we don't have transient state yet, but in the future we will have

Eikix commented on crates/contracts/src/kakarot_core/kakarot.cairo:
```diff
@@ -295,5 +301,38 @@ mod KakarotCore {
+                :evm_address,
+                :bytecode,
+                calldata: array![].span(),
+                :value,
```
as I understand it, execute does not perform value transfer yet,
i wonder if the value transfer flow will happen inside `execute` or somewhere else using our greater struct Journal or LocalStorage

enitrat replied:
Happen inside execute normally

Eikix replied:
ok

Eikix commented on crates/contracts/src/kakarot_core/kakarot.cairo:
```diff
@@ -295,5 +301,38 @@ mod KakarotCore {
+                return (ca.starknet_address, false);
+            }
+
+            ca.store_bytecode(execution_result.return_data);
```
this is an operation that can return an error, we should handle it

Eikix commented on crates/contracts/src/kakarot_core/kakarot.cairo:
```diff
@@ -295,5 +301,38 @@ mod KakarotCore {
+        /// # Returns
+        /// * The address the contract is deployed at - which is KakarotCore
+        /// * `true` if the operation is successfull, false otherwise
+        fn deploy_contract_account(
```
it may be worth it to look at how @ClementWalter does this in[ his recent PR](https://github.com/kkrt-labs/kakarot/pull/771/files#diff-10932ea6303303a44b4724d0a66ed9e547d2ea924e32f69ad1fbdcd4e68f9028R92)

Eikix commented on crates/evm/src/errors.cairo:
```diff
@@ -34,8 +38,27 @@ enum EVMError {
 
+#[generate_trait]
+impl EVMErrorImpl of EVMErrorTrait {
+    fn to_string(self: EVMError) -> felt252 {
```
Maybe we should altogether delete the Into<u256> I created, I'm not sure it's actually useful

Eikix commented on crates/evm/src/model/contract_account.cairo:
```diff
@@ -24,12 +27,48 @@ struct ContractAccount {
+    /// Returns an `EVMError` if the nonce of the contract account for
+    /// this EVM address was not `0`.
+    fn deploy(evm_address: EthAddress) -> Result<ContractAccount, EVMError> {
+        let mut maybe_ca = ContractAccountTrait::at(evm_address)?;
```
Should we actually check for account_at -> we don't want an EOA to be deployed there

ideally, we'd want to have stronger invariants, such as the inability to deploy an EOA if you don't own the private key to it

enitrat replied:
agreed

Eikix commented:
some conflicts

enitrat commented:
Related: #465

enitrat commented:
- Added some _very_ basic necessary error management features - these will most likely need to be refactored down the line to a stronger model
- Changed `decode_and_execute` to return if `pc>= bytecode_len` and set the status to Stopped
- use deploy in contract_account instead of direct access tests
- moved `deploy_contract_account` to `ContractAccountTrait::deploy` entirely. The KakarotCore contract doesn't need an internal function for this. #469

Eikix commented:
Some conflicts, reviewing now

Eikix commented on crates/evm/src/interpreter.cairo:
```diff
@@ -72,7 +72,8 @@ impl EVMInterpreterImpl of EVMInterpreterTrait {
         // Check if PC is not out of bounds.
         if pc >= bytecode_len {
-            return Result::Err(EVMError::InvalidProgramCounter(PC_OUT_OF_BOUNDS));
+            machine.set_stopped();
```
Already started to tackle this in #459 
But I think if pc == bytecode_len we stop,
if pc > bytecode_len, there's a problem, we should return an error,
though this should never happen, as jump forbids non-valid jumps

enitrat replied:
Let's get #459 merged before this PR then, as it's required to properly test

Eikix commented on crates/evm/src/interpreter.cairo:
```diff
@@ -72,7 +72,8 @@ impl EVMInterpreterImpl of EVMInterpreterTrait {
         if pc >= bytecode_len {
-            return Result::Err(EVMError::InvalidProgramCounter(PC_OUT_OF_BOUNDS));
+            machine.set_stopped();
+            return Result::Ok(());
```
also, as in #459 if pc == bytecode_len, then we need to wipe current_ctx.return_data, otherwise as part of #458, it'll get copied onto its parent return_data field

Eikix commented on crates/evm/src/machine.cairo:
```diff
@@ -16,7 +17,8 @@ struct Machine {
     memory: Memory,
-    storage_journal: Journal
+    storage_journal: Journal,
+    error: Option<EVMError>
```
we can revisit this later, I wonder how EELS or Geth bubbles up the error message or revert reason to the tx root ctx

Eikix replied:
Ahh interesting, EELS has an optional error field on their evm object:) lgtm

Eikix commented on crates/evm/src/machine.cairo:
```diff
@@ -235,6 +239,12 @@ impl MachineCurrentContextImpl of MachineCurrentContextTrait {
 
+
+    #[inline(always)]
+    fn error(ref self: Machine) -> Option<EVMError> {
```
must we consume machine here?
we can't go with a snapshot?

Eikix commented on crates/evm/src/model/contract_account.cairo:
```diff
@@ -24,12 +32,62 @@ struct ContractAccount {
+        origin: EthAddress, evm_address: EthAddress, value: u256, bytecode: Span<u8>
+    ) -> Result<ContractAccount, EVMError> {
+        let mut maybe_acc = AccountTrait::account_at(evm_address)?;
+        match maybe_acc {
```
I've seen this pattern a few times in this PR,
why not a simple early return inside an if?

```
let maybe_acc = AccountTrait::account_at(evm_address)?;
if maybe_acc.is_some() {
return ...
}
```

enitrat replied:
yep

Eikix commented on crates/evm/src/model/contract_account.cairo:
```diff
@@ -24,12 +32,62 @@ struct ContractAccount {
+            Option::Some(acc) => { return Result::Err(EVMError::DeployError(ACCOUNT_EXISTS)); },
+            Option::None => {}
+        }
+        let execution_result = execute(
```
Are we sure this is the responsibiliy of AccountTrait::deploy to know of the EVM?
I'd argue AccountTrait::deploy should come after the execute is successful, and execute should be called outside of it with the proper arguments

Eikix replied:
Or at least we call it process_create / process_create_message like the executions-specs?

enitrat replied:
> I'd argue AccountTrait::deploy should come after the execute is successful, and execute should be called outside

I can agree with that indeed - question is, where should we do this `execute` then. Revert back to how it was inside KakarotCore?

enitrat commented:
Restricting the scope of this PR, will split in more issues

https://github.com/kkrt-labs/kakarot-ssj/issues/474

enitrat commented on crates/evm/src/tests/test_model.cairo:
```diff
@@ -25,16 +27,17 @@ fn test_account_at_eoa() {
     let native_token = deploy_native_token();
     let kakarot_core = deploy_kakarot_core(native_token.contract_address);
-    // kakarot_core.deploy_contrat_account(...);
+    // We need to set_contract_address as the nonce is stored inside the contract that calls this function - here, it's the test contract by default.
```
funny thing heh

Eikix reviewed (approved):
lgtm
